tree/ntuple/v7/test/ntuple_basics.cxx

```diff
@@ -0,0 +1,345 @@
+
+   auto modelWrite = RNTupleModel::Create();
+   auto wrPt = modelWrite->MakeField<float>("pt", 42.0);
+   auto wrTag = modelWrite->MakeField<std::string>("tag", "xyz");
+   auto wrNnlo = modelWrite->MakeField<std::vector<std::vector<float>>>("nnlo");
+   auto wrFourVec = modelWrite->MakeField<std::array<float, 4>>("fourVec");
+   wrNnlo->push_back(std::vector<float>());
+   wrNnlo->push_back(std::vector<float>{1.0});
+   wrNnlo->push_back(std::vector<float>{1.0, 2.0, 4.0, 8.0});
+   wrFourVec->at(0) = 0.0;
+   wrFourVec->at(1) = 1.0;
+   wrFourVec->at(2) = 2.0;
+   wrFourVec->at(3) = 3.0;
+
+   auto modelRead = std::unique_ptr<RNTupleModel>(modelWrite->Clone());
+
+   {
+      RNTupleWriter ntuple(std::move(modelWrite),
+         std::make_unique<RPageSinkFile>("myNTuple", fileGuard.GetPath(), RNTupleWriteOptions()));
+      ntuple.Fill();
+      ntuple.CommitCluster();
+      *wrPt = 24.0;
+      wrNnlo->clear();
+      *wrTag = "";
+      wrFourVec->at(2) = 42.0;
+      ntuple.Fill();
+      *wrPt = 12.0;
+      wrNnlo->push_back(std::vector<float>{42.0});
+      *wrTag = "12345";
+      wrFourVec->at(1) = 24.0;
+      ntuple.Fill();
+   }
+
+   auto rdPt = modelRead->Get<float>("pt");
+   auto rdTag = modelRead->Get<std::string>("tag");
+   auto rdNnlo = modelRead->Get<std::vector<std::vector<float>>>("nnlo");
+   auto rdFourVec = modelRead->Get<std::array<float, 4>>("fourVec");
+
+   RNTupleReader ntuple(std::move(modelRead),
+      std::make_unique<RPageSourceFile>("myNTuple", fileGuard.GetPath(), RNTupleReadOptions()));
+   EXPECT_EQ(3U, ntuple.GetNEntries());
+
+   ntuple.LoadEntry(0);
+   EXPECT_EQ(42.0, *rdPt);
+   EXPECT_STREQ("xyz", rdTag->c_str());
+   EXPECT_EQ(3U, rdNnlo->size());
+   EXPECT_EQ(0U, (*rdNnlo)[0].size());
+   EXPECT_EQ(1U, (*rdNnlo)[1].size());
+   EXPECT_EQ(4U, (*rdNnlo)[2].size());
+   EXPECT_EQ(1.0, (*rdNnlo)[1][0]);
+   EXPECT_EQ(1.0, (*rdNnlo)[2][0]);
+   EXPECT_EQ(2.0, (*rdNnlo)[2][1]);
+   EXPECT_EQ(4.0, (*rdNnlo)[2][2]);
+   EXPECT_EQ(8.0, (*rdNnlo)[2][3]);
+   EXPECT_EQ(0.0, (*rdFourVec)[0]);
+   EXPECT_EQ(1.0, (*rdFourVec)[1]);
+   EXPECT_EQ(2.0, (*rdFourVec)[2]);
+   EXPECT_EQ(3.0, (*rdFourVec)[3]);
+
+   ntuple.LoadEntry(1);
+   EXPECT_EQ(24.0, *rdPt);
+   EXPECT_STREQ("", rdTag->c_str());
+   EXPECT_TRUE(rdNnlo->empty());
+   EXPECT_EQ(42.0, (*rdFourVec)[2]);
+
+   ntuple.LoadEntry(2);
+   EXPECT_EQ(12.0, *rdPt);
+   EXPECT_STREQ("12345", rdTag->c_str());
+   EXPECT_EQ(1U, rdNnlo->size());
+   EXPECT_EQ(1U, (*rdNnlo)[0].size());
+   EXPECT_EQ(42.0, (*rdNnlo)[0][0]);
+   EXPECT_EQ(24.0, (*rdFourVec)[1]);
+}
+
+TEST(RNTuple, TypeName) {
+   EXPECT_STREQ("float", ROOT::Experimental::RField<float>::TypeName().c_str());
+   EXPECT_STREQ("std::vector<std::string>",
+                ROOT::Experimental::RField<std::vector<std::string>>::TypeName().c_str());
+   EXPECT_STREQ("CustomStruct",
+                ROOT::Experimental::RField<CustomStruct>::TypeName().c_str());
+}
+
+TEST(RNTuple, Descriptor)
```

Comment:
The descriptor can become a separate collection I think.

Reply:
Makes sense, done in 21d0688